### ISSUE TITLE
Inference on trained model does not reflect test or train output during model training

### ISSUE THREAD

**Issue by dboshardy** [yuval-alaluf/restyle-encoder#72]
After training an image translation model (similar to toonify translation of real face to toonified face using a blended network) with paired images, the train output images look good. The test ones are hit and miss.  When using the best_model or the latest model with inference, both output essentially the latent average, not the appropriate translated image (which I have obtained via optimization via ffhq model and then passing the latent through the blended network.

I've tried more iterations per batch when running inference, but that doesn't seem to really change the output.  Justin Pinkney used pix2pixHD to do the direct image translation for his toonification example.  Does that approach not work using restyle-psp instead of pix2pixHD?

For inference, i'm using scripts/infererence_iterative_save_coupled.py and scripts/inference_iterative.py

**Comment by dboshardy** [comment]
I'll reiterate. the model seems to perform the translation very well given the train and test images while training.  The issue is when inferring via the checkpoint that generated such good results.

**Comment by yuval-alaluf** [comment]
If I understood correctly, the images in the `logs` directory of the experiment seem good, but when you run inference the results are like the latent average? If you take an image from the train set that looked good during training and tried running it with the `infererence_iterative_save_coupled.py` script, what do you get? 
If all the results look like the latent average, I would think that the checkpoint wasn't properly loaded in the inference script. Any chance you can provide a few examples of the results during training (on a test image) and what you get from the inference script?

**Comment by dboshardy** [comment]
Looking at some test images side-by-side, the poses change a bit to reflect the input image, but the rest of the features are pretty much the latent average with differences in lighting.

Unfortunately I'm unable to share any of the actual images with you at this time.

**Comment by yuval-alaluf** [comment]
> I did try running a test image through that was in the test set, and the resulting image is not what I see in the test images. I tried diffing the various checkpoints and they do differ.

Does this mean that if you take an image that looks good in the `logs` directory of your training directory and run it with the inference script you get different results? This shouldn't happen and either means that the checkpoint was not loaded correctly. Did you verify that all the weights were loaded correctly in the inference script?

**Comment by dboshardy** [comment]
I'll give that a go. Thank you!

**Comment by dboshardy** [comment]
Running a later model (I left it training) it produces the same result, no error thrown loading the model weights.  

I'm looking through the test/train outputs while training.  The train outputs are significantly better than the test outputs. Is that expected?  I admit, I should have been paying attention to both.

Tangent: Is there a way to resume training from a previous checkpoint?

**Comment by yuval-alaluf** [comment]
> Perhaps the ffhq model I fine-tuned and blended differs from the one used to train the restyle_psp_encoder?

This is definitely possible. 

> The train outputs are significantly better than the test outputs. Is that expected?

Because you are working with paired data, the loss weights may differ from what we used (since our setting was an unpaired setting). I am also unsure how many images you have so maybe there's a chance at overfitting? 
Since you are unable to share the images, my recommendation is to try to play a bit with the loss parameters to try to assist the model in generalizing to the test images. 

You may also want to consider training for a low number of iterative loops (`n_iters_per_batch`) to see if that may have an effect on the results. When inverting images, you want this to be around 5 to gradually get better results, but if you're performing image-to-image translation, maybe you'd be better off setting this to 2, for example.

**Comment by dboshardy** [comment]
Lots of things to try! Thank you so much for your input!

**Comment by dboshardy** [comment]
@yuval-alaluf for my own edification, why would lowering the number of iterations work better for image-to-image translation?

I'm trying to get a sense of how each of the training flags might affect training in different ways.
